In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('MERGED2017_18_PP.CSV', delimiter = ',')

/Users/williamegan/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (31,1726,1727,1728,1729,1815,1817,1823,1830,1879,1880,1881,1882,1883,1884,1885,1886,1887,1888,1889,1890,1891,1892,1893,1894,1895,1896,1897,1898,1957,1958,1959,1960,1961,1962,1963,1964,1965,1966,1967,1968,1969,1970,1971,1972,1973,1974,1975,1976) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
data.head()

UNITID    OPEID  OPEID6                               INSTNM        CITY  \
0  100654   100200    1002             Alabama A & M University      Normal   
1  100663   105200    1052  University of Alabama at Birmingham  Birmingham   
2  100690  2503400   25034                   Amridge University  Montgomery   
3  100706   105500    1055  University of Alabama in Huntsville  Huntsville   
4  100724   100500    1005             Alabama State University  Montgomery   

  STABBR         ZIP                                       ACCREDAGENCY  \
0     AL       35762  Southern Association of Colleges and Schools C...   
1     AL  35294-0110  Southern Association of Colleges and Schools C...   
2     AL  36117-3553  Southern Association of Colleges and Schools C...   
3     AL       35899  Southern Association of Colleges and Schools C...   
4     AL  36104-0271  Southern Association of Colleges and Schools C...   

                     INSTURL  \
0              www.aamu.edu/   
1                www.uab.edu   
2  www.amridgeuniversity.edu   
3                www.uah.edu   
4              www.alasu.edu   

                                              NPCURL  ...  \
0  https://galileo.aamu.edu/NetPriceCalculator/np...  ...   
1           uab.studentaidcalculator.com/survey.aspx  ...   
2                   www2.amridgeuniversity.edu:9091/  ...   
3                                    finaid.uah.edu/  ...   
4  www.alasu.edu/cost-aid/forms/calculator/index....  ...   

   OMAWDP8_NOTFIRSTTIME_POOLED_SUPP  OMENRUP_NOTFIRSTTIME_POOLED_SUPP  \
0                            0.4048                            0.2645   
1                            0.5448                              0.12   
2                            0.4647                            0.1912   
3                            0.5121                            0.1593   
4                            0.4226                            0.0754   

   OMENRYP_FULLTIME_POOLED_SUPP  OMENRAP_FULLTIME_POOLED_SUPP  \
0                        0.0137                         0.315   
1                        0.0728                        0.2583   
2                        0.0148                        0.3284   
3                        0.0165                         0.323   
4                        0.0254                        0.5327   

   OMAWDP8_FULLTIME_POOLED_SUPP  OMENRUP_FULLTIME_POOLED_SUPP  \
0                        0.3755                        0.2958   
1                        0.5872                        0.0817   
2                        0.4391                        0.2177   
3                        0.5361                        0.1243   
4                        0.3133                        0.1286   

   OMENRYP_PARTTIME_POOLED_SUPP  OMENRAP_PARTTIME_POOLED_SUPP  \
0                        0.0289                        0.4224   
1                        0.1398                        0.2618   
2                         0.028                        0.3645   
3                        0.0245                        0.3947   
4                         0.021                        0.5804   

   OMAWDP8_PARTTIME_POOLED_SUPP  OMENRUP_PARTTIME_POOLED_SUPP  
0                        0.2635                        0.2852  
1                        0.3977                        0.2008  
2                        0.4486                        0.1589  
3                        0.3476                        0.2331  
4                        0.1678                        0.2308  

[5 rows x 1977 columns]

In [4]:
data.columns

Index(['UNITID', 'OPEID', 'OPEID6', 'INSTNM', 'CITY', 'STABBR', 'ZIP',
       'ACCREDAGENCY', 'INSTURL', 'NPCURL',
       ...
       'OMAWDP8_NOTFIRSTTIME_POOLED_SUPP', 'OMENRUP_NOTFIRSTTIME_POOLED_SUPP',
       'OMENRYP_FULLTIME_POOLED_SUPP', 'OMENRAP_FULLTIME_POOLED_SUPP',
       'OMAWDP8_FULLTIME_POOLED_SUPP', 'OMENRUP_FULLTIME_POOLED_SUPP',
       'OMENRYP_PARTTIME_POOLED_SUPP', 'OMENRAP_PARTTIME_POOLED_SUPP',
       'OMAWDP8_PARTTIME_POOLED_SUPP', 'OMENRUP_PARTTIME_POOLED_SUPP'],
      dtype='object', length=1977)

In [10]:
null_dict = {}
for col in data.columns:
    null_dict[col] = 0
    for e in data[col]:
        if pd.isnull(e):
            null_dict[col] += 1

Now we are going to try and remove most of these columns. 

We will do this by looking at each value in our dictionary of the columns and each of their total number of null values, and only returning the ones that are have below a certain threshold of null values. We'll start with 3500, about 50%.


In [55]:
num_col = 0
cols_half = []
for col in null_dict:
    if null_dict[col] < 3500:
        num_col+=1
        cols_half.append(col)
        
num_col

395

In [56]:
cols_half

['UNITID',
 'OPEID',
 'OPEID6',
 'INSTNM',
 'CITY',
 'STABBR',
 'ZIP',
 'ACCREDAGENCY',
 'INSTURL',
 'NPCURL',
 'HCM2',
 'MAIN',
 'NUMBRANCH',
 'PREDDEG',
 'HIGHDEG',
 'CONTROL',
 'ST_FIPS',
 'REGION',
 'LOCALE',
 'LATITUDE',
 'LONGITUDE',
 'CCBASIC',
 'CCUGPROF',
 'CCSIZSET',
 'HBCU',
 'PBI',
 'ANNHI',
 'TRIBAL',
 'AANAPII',
 'HSI',
 'NANTI',
 'MENONLY',
 'WOMENONLY',
 'PCIP01',
 'PCIP03',
 'PCIP04',
 'PCIP05',
 'PCIP09',
 'PCIP10',
 'PCIP11',
 'PCIP12',
 'PCIP13',
 'PCIP14',
 'PCIP15',
 'PCIP16',
 'PCIP19',
 'PCIP22',
 'PCIP23',
 'PCIP24',
 'PCIP25',
 'PCIP26',
 'PCIP27',
 'PCIP29',
 'PCIP30',
 'PCIP31',
 'PCIP38',
 'PCIP39',
 'PCIP40',
 'PCIP41',
 'PCIP42',
 'PCIP43',
 'PCIP44',
 'PCIP45',
 'PCIP46',
 'PCIP47',
 'PCIP48',
 'PCIP49',
 'PCIP50',
 'PCIP51',
 'PCIP52',
 'PCIP54',
 'CIP01CERT1',
 'CIP01CERT2',
 'CIP01ASSOC',
 'CIP01CERT4',
 'CIP01BACHL',
 'CIP03CERT1',
 'CIP03CERT2',
 'CIP03ASSOC',
 'CIP03CERT4',
 'CIP03BACHL',
 'CIP04CERT1',
 'CIP04CERT2',
 'CIP04ASSOC',
 'CIP04CERT4',


In [58]:
data[cols_half].to_csv('2017_less_than_half_null.csv')

Even though we could go further with this, I kind of like this result. Going further shows that we lose some interesting things, and while there are still many columns here we may still want to remove, doing so will be much more manageable with a total of 395 columns vs. 1977.

In [50]:
num_col2 = 0
for col in null_dict:
    if null_dict[col] < 446:
        #we dont have any changes between 50 and 444, then small rises, then a 200 jump at 449 
        #This jump seems to be entirely CIP certifications
        num_col2+=1
        
num_col2

30

In [53]:
cols_below = []
for col in null_dict:
    if null_dict[col] < 446: 
        cols_below.append(col)
        
cols_below

['UNITID',
 'OPEID',
 'OPEID6',
 'INSTNM',
 'CITY',
 'STABBR',
 'ZIP',
 'INSTURL',
 'HCM2',
 'MAIN',
 'NUMBRANCH',
 'PREDDEG',
 'HIGHDEG',
 'CONTROL',
 'ST_FIPS',
 'REGION',
 'LOCALE',
 'LATITUDE',
 'LONGITUDE',
 'CCBASIC',
 'CCUGPROF',
 'CCSIZSET',
 'HSI',
 'MENONLY',
 'WOMENONLY',
 'DISTANCEONLY',
 'CURROPER',
 'ICLEVEL',
 'T4APPROVALDATE',
 'OPEFLAG']

In [54]:
data[cols_below]

UNITID    OPEID  OPEID6  \
0       100654   100200    1002   
1       100663   105200    1052   
2       100690  2503400   25034   
3       100706   105500    1055   
4       100724   100500    1005   
5       100751   105100    1051   
6       100760   100700    1007   
7       100812   100800    1008   
8       100830   831000    8310   
9       100858   100900    1009   
10      100937   101200    1012   
11      101028  1218200   12182   
12      101073  1055400   10554   
13      101116  1303906   13039   
14      101143   101500    1015   
15      101161   106000    1060   
16      101189   100300    1003   
17      101240   101700    1017   
18      101277  4187200   41872   
19      101286   101800    1018   
20      101295   787100    7871   
21      101301   569900    5699   
22      101365   962107    9621   
23      101435   101900    1019   
24      101453  2199700   21997   
25      101462   526000    5260   
26      101471  3002500   30025   
27      101480   102000    1020   
28      101499   106010    1060   
29      101505   102200    1022   
...        ...      ...     ...   
7082  47503301  4208601   42086   
7083  47503302  4208602   42086   
7084  47503303  4208605   42086   
7085  47503304  4208606   42086   
7086  47503305  4208603   42086   
7087  47503306  4208604   42086   
7088  47503307  4208607   42086   
7089  47503308  4208608   42086   
7090  47503309  4208609   42086   
7091  47503310  4208610   42086   
7092  47503311  4208611   42086   
7093  47503312  4208612   42086   
7094  47503313  4208613   42086   
7095  47647001  4088301   40883   
7096  47662901  4189101   41891   
7097  47691101  4205801   42058   
7098  48065701   869423    8694   
7099  48079001  4218901   42189   
7100  48092101  4220001   42200   
7101  48133101  4213501   42135   
7102  48154401  4220901   42209   
7103  48305401  4224601   42246   
7104  48312401   108131    1081   
7105  48312402   108161    1081   
7106  48312403   108162    1081   
7107  48312404   108163    1081   
7108  48387801  4223701   42237   
7109  48400201  4228101   42281   
7110  48778201  4221501   42215   
7111  49005401  4182601   41826   

                                                 INSTNM              CITY  \
0                              Alabama A & M University            Normal   
1                   University of Alabama at Birmingham        Birmingham   
2                                    Amridge University        Montgomery   
3                   University of Alabama in Huntsville        Huntsville   
4                              Alabama State University        Montgomery   
5                             The University of Alabama        Tuscaloosa   
6                     Central Alabama Community College    Alexander City   
7                               Athens State University            Athens   
8                       Auburn University at Montgomery        Montgomery   
9                                     Auburn University            Auburn   
10                          Birmingham Southern College        Birmingham   
11               Chattahoochee Valley Community College       Phenix City   
12                            Concordia College Alabama             Selma   
13                          South University-Montgomery        Montgomery   
14                   Enterprise State Community College        Enterprise   
15                    Coastal Alabama Community College       Bay Minette   
16                                  Faulkner University        Montgomery   
17                      Gadsden State Community College           Gadsden   
18                 New Beginning College of Cosmetology       Albertville   
19            George C Wallace Community College-Dothan            Dothan   
20    George C Wallace State Community College-Hance...        Hanceville   
21       George C Wallace State Community College-Selma             Selma   
22                        Herzing University-Birmingham        Bir